File: A2Z_sheet/Array/Easy/06_rotate_by_k/solution.js
var rotate = function (nums, k) {
	k %= nums.length;
	reverse(nums, 0, nums.length - 1);
	reverse(nums, 0, k - 1);
	reverse(nums, k, nums.length - 1);
};

var reverse = function (nums, start, end) {
	while (start < end) {
		[nums[start], nums[end]] = [nums[end], nums[start]];
		start++;
		end--;
	}
};

File: A2Z_sheet/Array/Easy/06_rotate_by_k/solution.rs
impl Solution {
    pub fn rotate(nums: &mut Vec<i32>, k: i32) {
        let n = nums.len() as i32;
        let k = k % n;
        Self::reverse(nums, 0, n - 1);
        Self::reverse(nums, 0, k - 1);
        Self::reverse(nums, k, n - 1);
    }

    fn reverse(nums: &mut Vec<i32>, start: i32, end: i32) {
        let mut start = start;
        let mut end = end;
        while start < end {
            nums.swap(start as usize, end as usize);
            start += 1;
            end -= 1;
        }
    }
}